src/viam/sdk/rpc/dial.cpp

```diff
@@ -72,18 +91,53 @@ const boost::optional<Credentials>& DialOptions::credentials() const {
     return credentials_;
 }
 
+int DialOptions::initial_connection_attempts() const {
+    return initial_connection_attempts_;
+}
+
 const std::chrono::duration<float>& DialOptions::timeout() const {
     return timeout_;
 }
 
-void DialOptions::set_allow_insecure_downgrade(bool allow) {
+std::chrono::duration<float> DialOptions::initial_connection_attempt_timeout() const {
+    return initial_connection_attempt_timeout_;
+}
+
+DialOptions& DialOptions::set_allow_insecure_downgrade(bool allow) {
     allow_insecure_downgrade_ = allow;
+
+    return *this;
 }
 
 bool DialOptions::allows_insecure_downgrade() const {
     return allow_insecure_downgrade_;
 }
 
+std::shared_ptr<ViamChannel> ViamChannel::dial_initial(
+    const char* uri, const boost::optional<DialOptions>& options) {
+    DialOptions opts = options.get_value_or(DialOptions());
+    auto timeout = opts.timeout();
+    auto attempts_remaining = opts.initial_connection_attempts();
+    if (attempts_remaining == 0) {
+        attempts_remaining = -1;
+    }
+    opts.set_timeout(opts.initial_connection_attempt_timeout());
+
+    std::exception exception;
+    while (attempts_remaining > 0) {
+        try {
+            auto connection = dial(uri, opts);
+            opts.set_timeout(timeout);
+            return connection;
+        } catch (const std::exception& e) {
+            attempts_remaining -= 1;
+            exception = e;
```

Comment:
This is going to slice the exception, I think (i.e. only the `std::exception` part will be copied, but any subtype will be lost). I think you can just look at `attempts_remaining` inside the `catch` and `throw` if there are non left.